uaclient/version.py

```diff
@@ -29,8 +41,44 @@ def get_version():
     if os.path.exists(os.path.join(topdir, ".git")):
         cmd = ["git", "describe", "--abbrev=8", "--match=[0-9]*", "--long"]
         try:
-            out, _ = system.subp(cmd)
+            out, _ = subp(cmd)
             return out.strip()
-        except exceptions.ProcessExecutionError:
+        except ProcessExecutionError:
             pass
     return __VERSION__
+
+
+def get_last_known_candidate() -> Optional[str]:
+    # If we can't determine when the cache was updated for the last time,
+    # We always assume it was as recent as possible - thus `inf`.
+    last_apt_cache_update = get_apt_cache_time() or inf
+    if (
+        not os.path.exists(CANDIDATE_CACHE_PATH)
+        or os.stat(CANDIDATE_CACHE_PATH).st_mtime < last_apt_cache_update
+    ):
+        try:
+            policy = get_apt_cache_policy_for_package("ubuntu-advantage-tools")
+            match = re.search(CANDIDATE_REGEX, policy)
+            if match:
+                candidate_version = match.group("candidate")
+                os.makedirs(UAC_TMP_PATH, exist_ok=True)
+                with open(CANDIDATE_CACHE_PATH, "w") as f:
+                    f.write(candidate_version)
+                os.chmod(CANDIDATE_CACHE_PATH, 0o777)
+        except Exception:
+            pass
+
+    try:
+        with open(CANDIDATE_CACHE_PATH, "r") as f:
+            return f.read()
```

Comment:
Let's just strip the output in case a newline or whitespace slips in
```suggestion
            return f.read().strip()
```